cmd/kube-controller-manager/app/controllermanager.go

```diff
@@ -275,7 +275,8 @@ func Run(c *config.CompletedConfig, stopCh <-chan struct{}) error {
 				run(ctx, startSATokenController, initializersFunc)
 			},
 			OnStoppedLeading: func() {
-				klog.Fatalf("leaderelection lost")
+				klog.ErrorS(nil, "leaderelection lost")
+				os.Exit(255)
```

Comment:
How quickly do you want to exit here?

Originally I had implemented JSON output handling so that `ErrorS` flushes buffered info messages automatically. @serathius later [removed that](https://github.com/kubernetes/kubernetes/pull/107035) because flushing also did an fsync, which caused a performance issue.

If you want to ensure that info messages are written, you have to add `klog.Flush` before `os.Exit`, but only if a (probably small) delay is acceptable.

Reply:
I met this case several time.  
When it crash, i think  the currect message is enough. It's not need print all groutines stack.
>If you want to ensure that info messages are written, you have to add klog.Flush before os.Exit, but only if a (probably small) delay is acceptable.

Added.

Reply:
Could you take a look at the official migration guide here?
https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/migration-to-structured-logging.md#replacing-fatal-calls

If we really need a Flush, then we are to add flush everywhere considering how many `klog.Fatalf` we had previously.

@serathius your input is greatly appreciated.

Reply:
> Could you take a look at the official migration guide here? https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/migration-to-structured-logging.md#replacing-fatal-calls
> 
> If we really need a Flush, then we are to add flush everywhere considering how many `klog.Fatalf` we had previously.
> 
> @serathius your input is greatly appreciated.

Thanks. I fixed according to [#replacing-fatal-calls](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/migration-to-structured-logging.md#replacing-fatal-calls)

Reply:
> If we really need a Flush, then we are to add flush everywhere considering how many klog.Fatalf we had previously.

The question is whether you *want* a Flush. Fatalf flushes for 10 seconds before exiting:
https://github.com/kubernetes/klog/blob/d78dad3276f1f4139a6614b08b02e4678d2c3393/klog.go#L1108-L1111

The migration guide doesn't cover that aspect; I don't know whether that was intentional or an oversight.

Reply:
May be in this we  do not need to add Flush before exit .   If need, we add a new PR ?

Reply:
> will ask around to see if that is intentional, or do we need to ma

@jiahuif  @pohly  Any update ?

Reply:
We'll discuss it in the [WG Structured Logging meeting](https://github.com/kubernetes/community/tree/master/wg-structured-logging#meetings) tomorrow.

Reply:
> I met this case several time. When it crash, i think the currect message is enough. It doesn't need to print all groutine stacks.
> 

The klog.Exit* APIs do exactly that [1]. It also does the flushing. However, this is not mentioned in the migration guide and may not be best practise. 

[1]
https://github.com/kubernetes/klog/blob/d78dad3276f1f4139a6614b08b02e4678d2c3393/klog.go#L1729-L1734

Reply:
@pchan: in the structured logging call today you said that `klog.Fatal` doesn't flush. But looking at the code I see a `timeoutFlush(10 * time.Second)` call before each `os.Exit`. What differs is the logic around stack traces.
Agreed? 

Our conclusion in the structured logging call was that flushing for a short amount of time before `os.Exit` may be useful and would mirror the old behavior better.

Several options exist for a `klog.Fatal` replacement:
```
klog:
var ExitFlushTimeout = 10 * time.Second

user code:
klog.ErrorS(...)
klog.FlushWithTimeout(klog.ExitFlushTimeout)
os.Exit(255)
```

Or a combined call:
```
klog.Errors(...)
klog.FlushAndExit(klog.ExitFlushTimeout, 255)
```

By requiring a timeout value in the API, users are encouraged to consider the implications (wait or exit immediately?). `klog.ExitFlushTimeout` could serve as default.

The rationale for having support for this in klog is that klog knows whether flushing is necessary. If it isn't, it can exit immediately.

I have not listed an alternative which combines logging with exit. Adding that would lead to the proliferation of different logging calls that we have right now and that we want to get away from.

Reply:
I have filed https://github.com/kubernetes/klog/issues/301.

As we have some people here who are working on the migration and/or own code which currently uses `klog.Fatal`, can we perhaps use this thread here to decide about the desired solution? See proposals above.

Reply:
I'd like to point out one advantage of a combined `klog.FlushAndExit`: if we use that consistently everywhere instead of `os.Exit`, then at some point we can really ban the usage of that function outside of `main` (https://github.com/kubernetes/kubernetes/issues/102231#issuecomment-1019041587).

/cc @sanposhiho

Reply:
> I'd like to point out one advantage of a combined `klog.FlushAndExit`: if we use that consistently everywhere instead of `os.Exit`, then at some point we can really ban the usage of that function outside of `main` ([#102231 (comment)](https://github.com/kubernetes/kubernetes/issues/102231#issuecomment-1019041587)).
> 
> /cc @sanposhiho

Wait you PR merged! 
If we want to change to klog.FlushAndExit. We should change everywhere by another PRs.

Reply:
> @pchan: in the structured logging call today you said that `klog.Fatal` doesn't flush. But looking at the code I see a `timeoutFlush(10 * time.Second)` call before each `os.Exit`. What differs is the logic around stack traces. Agreed?

Yes agree, apologies for the oversight.

> 
> Our conclusion in the structured logging call was that flushing for a short amount of time before `os.Exit` may be useful and would mirror the old behavior better.

One alternative viewpoint:

Is there anyone who would actually use that feature i.e., specify a timeout ? Unless there is an actual requirement somewhere, it looks to me that we are introducing another lever, that will most likely be not used and taken as the default. In this thread the requirement is to not have the stack trace, the Exit APIs fits the bill. Have just 2 APIs, Exit which flushes/exits and Fatal that does a stack trace on top of what Exit does and call it out in the migration guilde.

Reply:
>  the Exit APIs fits the bill

You mean `klog.Exitf` and friends? That is part of the old-style structured logging API. We need a replacement for that when migrating to structured logging, just as we need a replacement for `klog.Fatalf`.

Reply:
> So in here, we should exit by `klog.FlushAndExit(klog.ExitFlushTimeout, 1)` ?

Yes, that would be good.

> May be we need a new tag in klog and replace by kubernetes ?

I'll get to that as soon as possible. There's still https://github.com/kubernetes/klog/pull/297 pending, then klog should be ready for a new release.

Reply:
> I'll get to that as soon as possible. There's still [kubernetes/klog#297](https://github.com/kubernetes/klog/pull/297) pending, then klog should be ready for a new release.

Look forward to it!

Reply:
See https://github.com/kubernetes/kubernetes/pull/108725